src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -993,6 +993,11 @@ private static SizeAndAlignment ComputeInstanceSize(MetadataType type, LayoutInt
                 }
```

Comment:
The `classLayoutSize` is only respected for value types by the runtime type loader: https://github.com/dotnet/runtime/blob/8c68710084b861c44b6868c0bd972ce24aed55dd/src/coreclr/vm/methodtablebuilder.cpp#L8832-L8835

```
if (type.IsValueType)
{
    if (classLayoutSize != 0)
    {
       instanceSize = LayoutInt.Max(instanceSize, new LayoutInt(classLayoutSize));
    }
    else
    {
        instanceSize = LayoutInt.AlignUp(instanceSize, alignment, target);
    }
}
```

Repro:

```
using System.Runtime.InteropServices;
using System.Runtime.CompilerServices;

for (;;)
{
    var start = GC.GetAllocatedBytesForCurrentThread();
    GC.KeepAlive(new MyClass());
    Console.WriteLine(GC.GetAllocatedBytesForCurrentThread() - start);
}

[StructLayout(LayoutKind.Explicit, Size = 1000)]
class MyClass
{
    [FieldOffset(0)]
    public byte[] a;
    [FieldOffset(8)]
    public byte b;
}
```

Prints 32 with regular CoreCLR and 1016 with native AOT.

Reply:
I feel like this should print `1000` on both runtimes instead? I don't think anything documents that oversizing is not supported for classes.

Reply:
The StructLayout and FieldOffset attributes were designed to control the marshalled unmanaged layout. `LayoutKind.Explicit` is fully respected for the marshalled layout. `Console.WriteLine(Marshal.SizeOf<MyClass>())` prints `1000` for my example as expected.

Later ExplicitLayout got overloaded to control overlapping of fields in the managed layout. This overloading introduced many non-sensical behaviors. User code should not depend on managed layout (except blittable structs) in the first place.